### PR TITLE
[ffmpeg] Fix cross-compile for android (#11522)

### LINKED ISSUE #11522: [ffmpeg:arm64-android] build failure
**Host Environment**
- OS: Windows 10 Pro 1909 build 18363.836
- Compiler: Visual Studio 2019 (ver 16.5.5)

**To Reproduce**
`./vcpkg install ffmpeg:arm64-android`

**Failure logs** 
```

Computing installation plan...
The following packages will be built and installed:
    ffmpeg[avresample,core]:arm64-android
Starting package 1/1: ffmpeg:arm64-android
Building package ffmpeg[avresample,core]:arm64-android...
-- Using cached D:/dev/vcpkg/downloads/ffmpeg-ffmpeg-n4.2.tar.gz
-- Using source at D:/dev/vcpkg/buildtrees/ffmpeg/src/n4.2-a4d149a5c6
-- Building Options: --enable-asm --enable-yasm --disable-doc --enable-debug --enable-runtime-cpudetect --disable-openssl --disable-ffmpeg --disable-ffplay --disable-ffprobe --disable-libvpx --disable-libx264 --disable-opencl --disable-lzma --disable-bzlib --enable-avresample --disable-cuda --disable-nvenc --disable-cuvid --disable-libnpp  --enable-cross-compile --target-os=win32 --arch=arm64
-- Building Release Options:
-- Building ffmpeg for Release
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:72 (message):
    Command failed: /bin/bash --noprofile --norc D:/dev/vcpkg/ports/ffmpeg/build_linux.sh D:/dev/vcpkg/buildtrees/ffmpeg/arm64-android-rel D:/dev/vcpkg/buildtrees/ffmpeg/src/n4.2-a4d149a5c6 D:/dev/vcpkg/packages/ffmpeg_arm64-android "--enable-asm --enable-yasm --disable-doc --enable-debug --enable-runtime-cpudetect --disable-openssl --disable-ffmpeg --disable-ffplay --disable-ffprobe --disable-libvpx --disable-libx264 --disable-opencl --disable-lzma --disable-bzlib --enable-avresample --disable-cuda --disable-nvenc --disable-cuvid --disable-libnpp  --enable-cross-compile --target-os=win32 --arch=arm64 "
    Working Directory: D:/dev/vcpkg/buildtrees/ffmpeg/arm64-android-rel
    Error code: The system cannot find the file specified
    See logs for more information:

Call Stack (most recent call first):
  ports/ffmpeg/portfile.cmake:202 (vcpkg_execute_required_process)
  scripts/ports.cmake:90 (include)


Error: Building package ffmpeg:arm64-android failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `.\vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: ffmpeg:arm64-android
  Vcpkg version: 2020.02.04-nohash-external

Additionally, attach any relevant sections from the log files above.
```

**Additional context**
Added manually arm64-android triplet for static library compilation
```
set(VCPKG_TARGET_ARCHITECTURE arm64)
set(VCPKG_CRT_LINKAGE dynamic)
set(VCPKG_LIBRARY_LINKAGE static)
set(VCPKG_CMAKE_SYSTEM_NAME Android)
```

### PR DESCRIPTION
**Pull request by Jackarain**
Fix #11522

### CODE REVIEW

**Comment by JackBoosY** [comment]
Can you please resolve the file conflicts?

Thanks.

**Comment by Jackarain** [comment]
> Can you please resolve the file conflicts?
> 
> Thanks.

Done

**Comment by JackBoosY** [comment]
cc @cenit for review this PR.

**Comment by cenit** [comment]
sorry i don't have much experience in android toolchain. i would test the PR and learn things along the road, so i am not sure i'd be very helpful right now.

File: ports/ffmpeg/portfile.cmake [hunk 1]
```diff
@@ -562,8 +638,6 @@ if (VCPKG_TARGET_IS_OSX)
     LIST(LENGTH VCPKG_OSX_ARCHITECTURES OSX_ARCH_COUNT)
 endif()
 
-set(OPTIONS_CROSS "")
-
```
Comment (JackBoosY) [inline ports/ffmpeg/portfile.cmake:566]: Please don't remove this.

**Review by JackBoosY — CHANGES_REQUESTED** [review]

File: ports/ffmpeg/portfile.cmake [hunk 2]
```diff
@@ -173,6 +173,60 @@ else()
 endif()
 
 set(OPTIONS "--enable-pic --disable-doc --enable-debug --enable-runtime-cpudetect")
+set(OPTIONS_CROSS "")
```
Comment (Jackarain) [inline ports/ffmpeg/portfile.cmake:176]: @JackBoosY 

Moved here

**Comment by Jackarain** [comment]
> sorry i don't have much experience in android toolchain. i would test the PR and learn things along the road, so i am not sure i'd be very helpful right now.

1. https://developer.android.com/ndk/downloads#stable-downloads download ndk and extract.
2. Set environment variable ANDROID_NDK_HOME to your ndk path.
3. vcpkg install ffmpeg:arm-android
4. ...

File: ports/ffmpeg/portfile.cmake [hunk 3]
```diff
@@ -173,6 +173,60 @@ else()
...
+    endif()
+
+    set(ANDROID_CC "${ANDROID_TOOLCHAIN}/bin/${ANDROID_TARGET}${ANDROID_API}-clang")
+    set(ANDROID_CXX "${ANDROID_TOOLCHAIN}/bin/${ANDROID_TARGET}${ANDROID_API}-clang++")
+    set(ANDROID_RANLIB "${ANDROID_TOOLCHAIN}/bin/llvm-ranlib")
+    set(ANDROID_AR "${ANDROID_TOOLCHAIN}/bin/llvm-ar")
+    set(ANDROID_AS "${ANDROID_CC}")
+    set(ANDROID_LD "${ANDROID_TOOLCHAIN}/bin/ld")
+    set(ANDROID_STRIP "${ANDROID_TOOLCHAIN}/bin/llvm-strip")
+    set(ANDROID_CROSS_PREFIX "${ANDROID_TOOLCHAIN}/bin/${ANDROID_TOOLS}-")
```
Comment (JackBoosY) [inline ports/ffmpeg/portfile.cmake:226]: Does the toolchain supports / export the required macros or flags?
I think we'd better don't use the toochain file.
See android toolchain [here](https://github.com/microsoft/vcpkg/blob/master/scripts/toolchains/android.cmake).

cc @talregev for review and give some suggestion.

**Review by JackBoosY — CHANGES_REQUESTED** [review]

File: ports/ffmpeg/portfile.cmake (hunk 3 above)
Comment (talregev) [reply ports/ffmpeg/portfile.cmake:226]: I am not sure how the `portfile.cmake` work. 
I know if you have flags on the toolchain [android.cmake](https://github.com/microsoft/vcpkg/blob/master/scripts/toolchains/android.cmake) you don't add on them in the triplet. 
The same principle is for `portfile.cmake`? If yes, it should remove from here. 
So all the duplicated flags should take from `android.cmake`
After deletion of duplication flags, I can review more easily the remain flags.

**Comment by cenit** [comment]
> > sorry i don't have much experience in android toolchain. i would test the PR and learn things along the road, so i am not sure i'd be very helpful right now.
> 
> 1. https://developer.android.com/ndk/downloads#stable-downloads download ndk and extract.
> 2. Set environment variable ANDROID_NDK_HOME to your ndk path.
> 3. vcpkg install ffmpeg:arm-android
> 4. ...

i know how to install android toolchain, but being able to review a PR requires a deeper knowledge…

File: ports/ffmpeg/portfile.cmake (hunk 3 above)
Comment (Jackarain) [reply ports/ffmpeg/portfile.cmake]: All right, now use the [android.cmake](https://github.com/microsoft/vcpkg/blob/master/scripts/toolchains/android.cmake)

File: ports/ffmpeg/portfile.cmake [hunk 4]
```diff
@@ -184,6 +185,47 @@ if(VCPKG_TARGET_ARCHITECTURE STREQUAL "x86" OR VCPKG_TARGET_ARCHITECTURE STREQUA
...
+
+    if(VCPKG_TARGET_ARCHITECTURE STREQUAL "arm")
+        set(ANDROID_CPU "armv7-a")
+    elseif(VCPKG_TARGET_ARCHITECTURE STREQUAL "arm64")
+        set(ANDROID_CPU "armv8-a")
+    elseif(VCPKG_TARGET_ARCHITECTURE STREQUAL "x86")
+        set(ANDROID_CPU "x86")
+    elseif(VCPKG_TARGET_ARCHITECTURE STREQUAL "x64")
+        set(ANDROID_CPU "x86-64")
+    endif()
```
Comment (talregev) [inline ports/ffmpeg/portfile.cmake:206]: These flags are look similar to android `android.cmake`. Also it not support armv6. It should take from triplet name, like we do in `android.cmake`.

Please take a look for triplet examples
https://github.com/BOINC/boinc/blob/master/3rdParty/vcpkg_ports/triplets/default/arm-android.cmake
https://github.com/BOINC/boinc/blob/master/3rdParty/vcpkg_ports/triplets/default/armv6-android.cmake

File: ports/ffmpeg/portfile.cmake [hunk 5]
```diff
@@ -184,6 +185,47 @@ if(VCPKG_TARGET_ARCHITECTURE STREQUAL "x86" OR VCPKG_TARGET_ARCHITECTURE STREQUA
...
+
+    set(ANDROID_API 21) # default ANDROID_API 21
+    if($ENV{ANDROID_API})
+        set(ANDROID_API $ENV{ANDROID_API})
+    endif()
```
Comment (talregev) [inline ports/ffmpeg/portfile.cmake:211]: It should taken from triplet and not environment. vcpkg team explain me how much env is bad. 
Please take a look on my example.

File: ports/ffmpeg/portfile.cmake [hunk 6]
```diff
@@ -184,6 +185,47 @@ if(VCPKG_TARGET_ARCHITECTURE STREQUAL "x86" OR VCPKG_TARGET_ARCHITECTURE STREQUA
...
+
+    if($ENV{ANDROID_PLATFORM})
+        set(ANDROID_PLATFORM $ENV{ANDROID_PLATFORM})
+    else()
+        set(ANDROID_PLATFORM ${ANDROID_API})
+    endif()
+
```
Comment (talregev) [inline ports/ffmpeg/portfile.cmake:218]: here is the same as `ANDROID_API`

**Review by talregev — CHANGES_REQUESTED** [review]

File: ports/ffmpeg/portfile.cmake (hunk 5 above)
Comment (Jackarain) [reply ports/ffmpeg/portfile.cmake:211]: I do not understand what you are saying.
Why is vcpkg so difficult to use, It's like shit.
Close this PR, it's boring.

Comment (Jackarain) [reply ports/ffmpeg/portfile.cmake:211]: @cenit deeper knowledge？hehe, when do you plan to fix this problem (#11522,) or decide not to fix it anymore?

**Pull request closed without merging**